### PR TITLE
CI: Add pre-commit for black, whitespaces, and end-of-line

### PR DESCRIPTION
**Pull request by math-fehr**
This adds support for `pre-commit` package, which can easily install pre-commit hooks.
The following hooks are added:
* Check for trailing whitespaces
* Check that there is a newline at the end of files
* Check for black formatting

In particular, this removes the black CI in favor of this one.

### CODE REVIEW

**Comment by math-fehr** [comment]
Perfect, the CI shows nicely what needs to be formatted

**Review by superlopuh — APPROVED** [review]

**Comment by superlopuh** [comment]
Do we want to format mlir and xdsl on pre-commit?

**Comment by superlopuh** [comment]
On second thought, I like it

**Comment by math-fehr** [comment]
I feel that removing whitespaces and fixing end-of-file newline is valuable in all files.
What do you think?

**Review by webmiche — COMMENTED** [review]
LGTM, it probably does make most sense to outsource this to people's setups instead of the CI.

Can we require that everyone installs the pre-commit hook before being able to commit to the repo? Otherwise, we will certainly get people that just did not install the hook becasue they did not read the readme :sweat_smile:

**Review by webmiche — APPROVED** [review]

File: .github/workflows/code-formatting.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  push:
+    branches: [main]
+
+jobs:
+  pre-commit:
+
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ['3.10', '3.11']
```
Comment (AntonLydike) [inline .github/workflows/code-formatting.yml:19]: Is this necessary?

**Review by AntonLydike — APPROVED** [review]
LGTM, just a minor nit

**Pull request merged**